src/index.js

```diff
@@ -29,365 +29,16 @@
-                                })
-                                .then(async device => {
-                                    const choices = Object.keys(dfu);
-                                    if (choices.length > 1 && promiseChoice) {
-                                        return { device, choice: await promiseChoice('Which firmware do you want to program?', choices) };
-                                    }
-                                    return { device, choice: choices.pop() };
-                                })
-                                .then(({ device, choice }) => (
-                                    prepareInDFUBootloader(device, dfu[choice])
-                                ))
-                                .then(resolve)
-                                .catch(reject);
-                        });
-                }
-                debug('Device is not in DFU-Bootloader and has no DFU trigger interface');
-            }
-        }
-
-        if (jprog && selectedDevice.traits.includes('jlink')) {
-            let firmwareFamily;
-            return verifySerialPortAvailable(selectedDevice)
-                .then(() => openJLink(selectedDevice))
-                .then(() => getDeviceFamily(selectedDevice))
-                .then(family => {
-                    firmwareFamily = jprog[family];
-                    if (!firmwareFamily) {
-                        throw new Error(`No firmware defined for ${family} family`);
-                    }
-                })
-                .then(() => validateFirmware(selectedDevice, firmwareFamily))
-                .then(valid => {
-                    if (valid) {
-                        debug('Applicaton firmware id matches');
-                        return selectedDevice;
-                    }
-                    return Promise.resolve()
-                        .then(async () => {
-                            if (!promiseConfirm) return;
-                            if (!await promiseConfirm('Device must be programmed, do you want to proceed?')) {
-                                throw new Error('Preparation cancelled by user');
-                            }
-                        })
-                        .then(() => programFirmware(selectedDevice, firmwareFamily));
-                })
-                .then(() => closeJLink(selectedDevice))
-                .then(() => resolve(selectedDevice))
-                .catch(reject);
-        }
-
-        debug('Selected device cannot be prepared, maybe the app still can use it');
-        return resolve(selectedDevice);
-    });
-}
-
-export default { setupDevice };
+import * as initPacket from './util/initPacket';
+import * as dfuTrigger from './dfuTrigger';
+import * as jprogFunc from './jprogFunc';
+import { setupDevice, waitForDevice, detachAndWaitFor } from './setupDevice';
+
+export default {
+    setupDevice,
+    waitForDevice,
+    detachAndWaitFor,
+    jprogFunc,
+    dfuTrigger,
+    initPacket,
```

Comment:
- `detachAndWaitFor` is the one call that you would more frequently do, so it's easier to use.
- `predictSerialNumberAfterReset` is looks like it's not necessary, can be removed, it was invented for an SDK functionality that is not yet implemented
- initPacket functionality is a complete copy from programmer, which is I think used by programmer

Reply:
Regarding grouping of DFU related functions is a good idea, however I didn't want to mix DFU and DFU trigger functionality, since they are quite different. How should it be?